src/array/cuda/csr_mm.cu

```diff
@@ -70,30 +70,109 @@ std::pair<CSRMatrix, NDArray> CusparseSpgemm(
       idtype, CUSPARSE_INDEX_BASE_ZERO, dtype));
   // SpGEMM Computation
   cusparseSpGEMMDescr_t spgemmDesc;
+  cusparseSpGEMMAlg_t alg = CUSPARSE_SPGEMM_DEFAULT;
+
   CUSPARSE_CALL(cusparseSpGEMM_createDescr(&spgemmDesc));
-  size_t workspace_size1 = 0, workspace_size2 = 0;
+  size_t workspace_size1 = 0, workspace_size2 = 0, workspace_size3 = 0;
   // ask bufferSize1 bytes for external memory
   CUSPARSE_CALL(cusparseSpGEMM_workEstimation(
       thr_entry->cusparse_handle, transA, transB, &alpha, matA, matB, &beta,
-      matC, dtype, CUSPARSE_SPGEMM_DEFAULT, spgemmDesc, &workspace_size1,
+      matC, dtype, alg, spgemmDesc, &workspace_size1,
       NULL));
   void* workspace1 = (device->AllocWorkspace(ctx, workspace_size1));
   // inspect the matrices A and B to understand the memory requiremnent
-  // for the next step
-  CUSPARSE_CALL(cusparseSpGEMM_workEstimation(
-      thr_entry->cusparse_handle, transA, transB, &alpha, matA, matB, &beta,
-      matC, dtype, CUSPARSE_SPGEMM_DEFAULT, spgemmDesc, &workspace_size1,
-      workspace1));
+  cusparseStatus_t e =
+    cusparseSpGEMM_workEstimation(thr_entry->cusparse_handle, transA,
+                                  transB, &alpha, matA, matB, &beta,
+                                  matC, dtype, alg, spgemmDesc,
+                                  &workspace_size1, workspace1);
+  // CUSPARSE_SPGEMM_DEFAULT not support getting num_prods > 2^31 -1
+  // and throws insufficient memory error within workEstimation call
+  if (e == CUSPARSE_STATUS_INSUFFICIENT_RESOURCES) {
+    // fall back to ALG2 to estimate num_prods
+    alg = CUSPARSE_SPGEMM_ALG2;
+    device->FreeWorkspace(ctx, workspace1);
+    // rerun cusparseSpGEMM_workEstimation
+    CUSPARSE_CALL(cusparseSpGEMM_workEstimation(thr_entry->cusparse_handle,
+                                                transA, transB, &alpha, matA,
+                                                matB, &beta, matC, dtype, alg,
+                                                spgemmDesc, &workspace_size1,
+                                                NULL));
+    void* workspace1 = (device->AllocWorkspace(ctx, workspace_size1));
```

Comment:
Yeah... Though it runs well with various problem sizes, I think it could result in memory leak and I should not define it again locally. I will make the change later. 
Thank you for catching it.